pilot/pkg/networking/apigen/apigen.go

```diff
@@ -0,0 +1,165 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package apigen
+
+import (
+	"strings"
+
+	gogotypes "github.com/gogo/protobuf/types"
+	golangany "github.com/golang/protobuf/ptypes/any"
+
+	mcp "istio.io/api/mcp/v1alpha1"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/serviceregistry/external"
+	"istio.io/istio/pkg/config/schema/collections"
+	"istio.io/istio/pkg/config/schema/resource"
+	"istio.io/pkg/log"
+)
+
+// APIGenerator supports generation of high-level API resources, similar with the MCP
+// protocol. This is a replacement for MCP, using XDS (and in future UDPA) as a transport.
+// Based on lessons from MCP, the protocol allows incremental updates by
+// default, using the same mechanism that EDS is using, i.e. sending only changed resources
+// in a push. Incremental deletes are sent as a resource with empty body.
+//
+// Example: networking.istio.io/v1alpha3/VirtualService
+//
+// TODO: we can also add a special marker in the header)
+type APIGenerator struct {
+}
+
+// TODO: take 'updates' into account, don't send pushes for resources that haven't changed
+// TODO: support WorkloadEntry - to generate endpoints (equivalent with EDS)
+// TODO: based on lessons from MCP, we want to send 'chunked' responses, like apiserver does.
+// A first attempt added a 'sync' record at the end. Based on feedback and common use, a
+// different approach can be used - for large responses, we can mark the last one as 'hasMore'
+// by adding a field to the envelope.
+
+// Generate implements the generate method for high level APIs, like Istio config types.
+// This provides similar functionality with MCP and :8080/debug/configz.
+//
+// Names are based on the current resource naming in istiod stores.
+func (g *APIGenerator) Generate(proxy *model.Proxy, push *model.PushContext, w *model.WatchedResource, updates model.XdsUpdates) model.Resources {
+	resp := []*golangany.Any{}
+
+	// Note: this is the style used by MCP and its config. Pilot is using 'Group/Version/Kind' as the
+	// key, which is similar.
+	//
+	// The actual type in the Any should be a real proto - which is based on the generated package name.
+	// For example: type is for Any is 'type.googlepis.com/istio.networking.v1alpha3.EnvoyFilter
+	// We use: networking.istio.io/v1alpha3/EnvoyFilter
```

Comment:
Confusing, which is the real one

Reply:
Both are real :-) 

TypeURL is the name we use in Store interface - based on k8s naming. 
The Resource Any must use the proto convention, so it can be unmarshalled with the standard library.